adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,117 @@
+
+== Decision Drivers
+
+* Anything we do should minimise work for extension authors
+* Anything we do should minimise breaking changes, especially since many quarkiverse extensions will be using back-level versions of the Quarkus extension plugin
+* We should think carefully about the right information layout, so that we get the right information in the right long-term place, and only need to do this disruptive re-arrangement once
+
+== Considered options
+
+- *Manual creation of metadata*. This is tedious for users and error-prone; we could not assume any manually-managed metadata was complete. It's also annoyingly non-DRY; we'd be asking extension owners to write the same information down in two places (their code, and a metadata file). It does have the advantage that it's non-breaking.
+- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
+- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
+- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment model, not the runtime module
+* The build of extensions' deployment module invokes a maven plugin (the existing one, or a new one)
+* (optional) Extension source code refactoring: `runtime/src/META-INF/quarkus-extension.yaml` moves to `deployment`. This eliminates one of the cross-module dependencies we'd like to avoid to stay cache-friendly.
```

Comment:
During the "transition period", we could consider looking for `quarkus-extension.yaml` "template" in the runtime module if it is not found in the deployment one when we build the deployment module.

Reply:
Yes, I agree. Does https://github.com/quarkusio/quarkus/pull/40336/files/f5e181616149a2e37573296f5caeaf8bb27a76ff#diff-6fae3f8e5c70cd771717b05a93172b01c4d2a52a47f3eef96e238890d1078386R73 already cover this? I could add the 'template' word in there to make it clearer.

Reply:
After the first time reading it, I thought it was about making the plugin recognize whether it's configured in the runtime or the deployment module and look for the YAML in that module whatever it is.
Perhaps, we could make more clear stating that if it is configured in the deployment module and the YAML template isn't found in the deployment module, it would look for it in the runtime module?